### PR TITLE
feat: add video option to splashscren

### PR DESCRIPTION
**Pull request by andrejborstnik**
Adds ability to use video as a splashscreen.
1. cannot show both image splash & video splash at same time
2. video is rendered to fit the screen, so it adapts for all screensizes, including tablets

## Test Plan

Follow instructions & setup with `exodus-mobile`. Test on phone & tablet, portrait & landscape mode

~~@mvayngrib how do you propose we merge this? Likely won't fit best on master, but I can try~~ compatible with master, will merge there & cherry-pick

### CODE REVIEW

File: android/src/main/java/org/devio/rn/splashscreen/SplashScreen.java [hunk 1]
```diff
@@ -23,13 +30,89 @@
...
+                    if (ratioWidth < ratioHeight ) {
+                        fullWidth = videoWidth * ratioWidth;
+                        fullHeight = videoHeight * ratioWidth;
+                    } else {
+                        fullWidth = videoWidth * ratioHeight;
+                        fullHeight = videoHeight * ratioHeight;
+                    }
+
+                    videoView.getLayoutParams().width = Math.round(fullWidth);
+                    videoView.getLayoutParams().height = Math.round(fullHeight);
```
Comment (andrejborstnik) [inline android/src/main/java/org/devio/rn/splashscreen/SplashScreen.java]: Need to set video width/height to preserve aspect ratio. If you know a better way please lmk

**Comment by alexandrius** [comment]
Don't you guys think using Lottie for the animation would be much better fit?

**Comment by andrejborstnik** [comment]
> Don't you guys think using Lottie for the animation would be much better fit?

Design had issues extracting some complex animations to lottie, that is why we use video. Might also be easier to use video controls - we will need to pause video in some cases

**Comment by alexandrius** [comment]
@andrejborstnik You can pause lottie animations https://github.com/lottie-react-native/lottie-react-native/blob/master/docs/api.md#methods-imperative-api
I have played with Bodymovin and have fixed few animations export to Lottie

**Comment by alexandrius** [comment]
~@andrejborstnik Why not use react-native-video for faking splash screen?~
Found answer https://github.com/ExodusMovement/exodus-mobile/pull/11240

File: android/src/main/java/org/devio/rn/splashscreen/SplashScreen.java [hunk 2]
```diff
@@ -23,13 +30,89 @@
...
+        if (isImageActive || isVideoActive) return;
+        isVideoActive = true;
+
+        mActivity = new WeakReference<Activity>(activity);
+        activity.runOnUiThread(new Runnable() {
+            @Override
+            public void run() {
+                if (!activity.isFinishing()) {
+                    mSplashDialog = new Dialog(activity, R.style.SplashScreen_Fullscreen);
+                    mSplashDialog.setContentView(R.layout.video_view);
```
Comment (alexandrius) [inline android/src/main/java/org/devio/rn/splashscreen/SplashScreen.java:48]: Can't find this xml

Comment (alexandrius) [reply android/src/main/java/org/devio/rn/splashscreen/SplashScreen.java:48]: Will the layout be part of Exodus app structure?

File: android/src/main/java/org/devio/rn/splashscreen/SplashScreen.java [hunk 3]
```diff
@@ -23,13 +30,89 @@
...
+                    float fullHeight;
+                    if (ratioWidth < ratioHeight ) {
+                        fullWidth = videoWidth * ratioWidth;
+                        fullHeight = videoHeight * ratioWidth;
+                    } else {
+                        fullWidth = videoWidth * ratioHeight;
+                        fullHeight = videoHeight * ratioHeight;
+                    }
+
+                    videoView.getLayoutParams().width = Math.round(fullWidth);
```
Comment (alexandrius) [inline android/src/main/java/org/devio/rn/splashscreen/SplashScreen.java]: Why not cast: `(int) fullWidth`

File: ios/RNSplashScreen.m [hunk 4]
```diff
@@ -20,9 +22,53 @@ - (dispatch_queue_t)methodQueue{
...
+
+  NSURL *videoURL = [NSURL fileURLWithPath:videoPath];
+  AVPlayer *player = [AVPlayer playerWithURL:videoURL];
+  playerLayer.videoGravity = AVLayerVideoGravityResizeAspectFill;
+
+  AVPlayerLayer *playerLayer = [AVPlayerLayer playerLayerWithPlayer:player];
+  playerLayer.frame = rootView.bounds;
+  [playerLayer setName:RNSplashScreenOverlayName];
+
+  [rootView.layer addSublayer:playerLayer];
```
Comment (alexandrius) [inline ios/RNSplashScreen.m:44]: Any reason to use layer instead of view?

**Comment by alexandrius** [comment]
The change is very distinct from original react-native-splash-screen. I also see links to files which are linked inside the Exodus app instead of this library. I think we should move this code into the project

File: android/src/main/java/org/devio/rn/splashscreen/SplashScreen.java (hunk 2 above)
Comment (andrejborstnik) [reply android/src/main/java/org/devio/rn/splashscreen/SplashScreen.java:48]: Yes, check updated docs. I added it in refs.xml

File: android/src/main/java/org/devio/rn/splashscreen/SplashScreen.java (hunk 3 above)
Comment (andrejborstnik) [reply android/src/main/java/org/devio/rn/splashscreen/SplashScreen.java]: No particular reason, will update

File: ios/RNSplashScreen.m (hunk 4 above)
Comment (andrejborstnik) [reply ios/RNSplashScreen.m:44]: Attaching view directly didn't seem to work. Maybe I was doing something wrong though

**Comment by andrejborstnik** [comment]
> The change is very distinct from original react-native-splash-screen. I also see links to files which are linked inside the Exodus app instead of this library. I think we should move this code into the project

The original also loads files from consuming project. In that sense this change follows existing patterns precisely. And overall seems like the same solution to me, just displays a video instead of image

File: android/src/main/java/org/devio/rn/splashscreen/SplashScreen.java (hunk 2 above)
Comment (alexandrius) [reply android/src/main/java/org/devio/rn/splashscreen/SplashScreen.java:48]: I think this library and project are very heavily interlinked. I think either we change the code to not rely on the app source or we should move it into the project.

File: ios/RNSplashScreen.m (hunk 4 above)
Comment (alexandrius) [reply ios/RNSplashScreen.m:44]: Layer is just fine, was curious.

**Comment by alexandrius** [comment]
> The original also loads files from consuming project. In that sense this change follows existing patterns precisely. And overall seems like the same solution to me, just displays a video instead of image

I see. The pattern used by this project doesn't feel very right to me tho.

File: android/src/main/java/org/devio/rn/splashscreen/SplashScreen.java (hunk 2 above)
Comment (andrejborstnik) [reply android/src/main/java/org/devio/rn/splashscreen/SplashScreen.java:48]: They interlinked, but no more than before. From the main project we take:
* video_view layout -> I guess we can move this to the project, don't see a downside
* video content -> it is expected this is a requirement. We could pass it in as an argument instead, but the amount of linkage would be the same (& we change existing pattern of the lib)

Comment (alexandrius) [reply android/src/main/java/org/devio/rn/splashscreen/SplashScreen.java:48]: Just checked. Didn’t know about it

**Comment by alexandrius** [comment]
utACK

**Review by alexandrius — APPROVED** [review]

File: android/src/main/java/org/devio/rn/splashscreen/SplashScreen.java (hunk 2 above)
Comment (andrejborstnik) [reply android/src/main/java/org/devio/rn/splashscreen/SplashScreen.java:48]: 9904972fea067890fb896172ce2ced0496d90b27

**Pull request merged**